#12717 Add a new section: automation of local dev environments

markgreene74 opened the pull request:
## References

This PR is related to #12717.

## Code changes

With this PR a new sub-section "Using automation to set up a local development environment" is added to "Setting up a local development environment" in `contributing.rst`.

The sub-section contains a small description of some of the advantages and link to a working example of Vagrantfile and additional documentation.

## User-facing changes

Changes to the "Contribute" documentation.

## Backwards-incompatible changes

N/A, no changes to JupyterLab public APIs.

jtpio commented:
Thanks @markgreene74 for taking the time to document this setup :+1: 

This PR was discussed during the July 20th weekly meeting: https://github.com/jupyterlab/team-compass/issues/152#issuecomment-1191170828

I think the overall feeling was that it can indeed be difficult to automate a development environment for JupyterLab. There are multiple technologies involves such as Python and JavaScript / TypeScript with Node.js. So any effort to make contributing to the project easier is really appreciated!

However it looked like no-one in the current JupyterLab team is familiar with Vagrant. So there is a risk the documentation will not be maintained or will become outdated.

I think there is still value in mentioning the Vagrant based setup in the contributing docs. Maybe we could simply link to your repository so the folks interested can find all the information there? https://github.com/markgreene74/jupyterlab-local-dev-with-vagrant
Then there could be a subsection in https://jupyterlab.readthedocs.io/en/latest/developer/contributing.html#setting-up-a-local-development-environment still called `Using automation to set up a local development environment`, but only keeping a small description of what it is about. For example:

> While there is a lot to learn by following the steps above, they can be automated to save time. This section shows how
to do that using Vagrant as an example.
> The main advantages of using automation are: reduced time to get the environment up-and-running, reduced time to
re-build the environment, better standardisation ("baseline", reproducible environments).
> A practical example can be found `here <https://github.com/markgreene74/jupyterlab-local-dev-with-vagrant>`_ and
includes a ``Vagrantfile``, the bootstrap files and additional documentation.

What do you think? Thanks again for working on this!

markgreene74 commented:
Hi @jtpio !

Thank you very much for the feedback. It makes perfectly sense, I will make the changes now :+1:

markgreene74 commented:
That's done now. I implemented the changes suggested (single sub-section), rebased to master and put everything in a single commit so hopefully should be easier to review.

Also, made the necessary changes to the PR description.

fcollonval commented on docs/source/developer/contributing.rst:
```diff
@@ -297,6 +297,18 @@ To check which version of Node.js is installed:
+The main advantages of using automation are: reduced time to get the environment up-and-running, reduced time to
+re-build the environment, better standardisation ("baseline", reproducible environments).
+
+A practical example can be found `here <https://github.com/markgreene74/jupyterlab-local-dev-with-vagrant>`_ and
```
let's use _there_ rather than _here_ to highlight it is outside of the main documentation.

```suggestion
A practical example can be found `there <https://github.com/markgreene74/jupyterlab-local-dev-with-vagrant>`_ and
```

markgreene74 replied:
Thank you @fcollonval , I hadn't thought of that.

fcollonval reviewed (approved):
Thanks a lot for the patience @markgreene74

markgreene74 commented:
Thank you for the review @fcollonval :)
